File: src/main/java/link/infra/packwiz/installer/metadata/hash/Murmur2Hasher.java
package link.infra.packwiz.installer.metadata.hash;

import okio.Buffer;
import okio.Source;

import java.io.IOException;

public class Murmur2Hasher implements IHasher {
	private class Murmur2GeneralHashingSource extends GeneralHashingSource {
		Murmur2Hash value;
		Buffer internalBuffer = new Buffer();
		Buffer tempBuffer = new Buffer();
		Source delegate;

		public Murmur2GeneralHashingSource(Source delegate) {
			super(delegate);
			this.delegate = delegate;
		}

		@Override
		public long read(Buffer sink, long byteCount) throws IOException {
			long out = delegate.read(tempBuffer, byteCount);
			if (out > -1) {
				sink.write(tempBuffer.clone(), out);
				internalBuffer.write(tempBuffer, out);
			}
			return out;
		}

		@Override
		public Hash getHash() {
			if (value == null) {
				byte[] data = computeNormalizedArray(internalBuffer.readByteArray());
				value = new Murmur2Hash(Murmur2Lib.hash32(data, data.length, 1));
			}
			return value;
		}

		// Credit to https://github.com/modmuss50/CAV2/blob/master/murmur.go
		private byte[] computeNormalizedArray(byte[] input) {
			byte[] output = new byte[input.length];
			int num = 0;
			for (byte b : input) {
				if (!(b == 9 || b == 10 || b == 13 || b == 32)) {
					output[num] = b;
					num++;
				}
			}
			byte[] outputTrimmed = new byte[num];
			System.arraycopy(output, 0, outputTrimmed, 0, num);
			return outputTrimmed;
		}

	}

	private static class Murmur2Hash extends Hash {
		int value;
		private Murmur2Hash(String value) {
			// Parsing as long then casting to int converts values gt int max value but lt uint max value
			// into negatives. I presume this is how the murmur2 code handles this.
			this.value = (int)Long.parseLong(value);
		}

		private Murmur2Hash(int value) {
			this.value = value;
		}

		@Override
		public boolean equals(Object obj) {
			if (!(obj instanceof Murmur2Hash)) {
				return false;
			}
			Murmur2Hash objHash = (Murmur2Hash) obj;
			return value == objHash.value;
		}

		@Override
		public String toString() {
			return "murmur2: " + value;
		}

		@Override
		protected String getStringValue() {
			return Integer.toString(value);
		}

		@Override
		protected String getType() {
			return "murmur2";
		}
	}

	@Override
	public GeneralHashingSource getHashingSource(Source delegate) {
		return new Murmur2GeneralHashingSource(delegate);
	}

	@Override
	public Hash getHash(String value) {
		return new Murmur2Hash(value);
	}
	
}

File: src/main/kotlin/link/infra/packwiz/installer/metadata/hash/Murmur2Hasher.kt
package link.infra.packwiz.installer.metadata.hash

import okio.Buffer
import okio.Source
import java.io.IOException

class Murmur2Hasher : IHasher {
	private inner class Murmur2GeneralHashingSource(delegate: Source) : GeneralHashingSource(delegate) {
		val internalBuffer = Buffer()
		val tempBuffer = Buffer()

		override val hash: Hash by lazy(LazyThreadSafetyMode.NONE) {
			val data = internalBuffer.readByteArray()
			Murmur2Hash(Murmur2Lib.hash32(data, data.size, 1))
		}

		@Throws(IOException::class)
		override fun read(sink: Buffer, byteCount: Long): Long {
			val out = delegate.read(tempBuffer, byteCount)
			if (out > -1) {
				sink.write(tempBuffer.clone(), out)
				computeNormalizedBufferFaster(tempBuffer, internalBuffer)
			}
			return out
		}

		// Credit to https://github.com/modmuss50/CAV2/blob/master/murmur.go
//		private fun computeNormalizedArray(input: ByteArray): ByteArray {
//			val output = ByteArray(input.size)
//			var index = 0
//			for (b in input) {
//				when (b) {
//					9.toByte(), 10.toByte(), 13.toByte(), 32.toByte() -> {}
//					else -> {
//						output[index] = b
//						index++
//					}
//				}
//			}
//			val outputTrimmed = ByteArray(index)
//			System.arraycopy(output, 0, outputTrimmed, 0, index)
//			return outputTrimmed
//		}

		private fun computeNormalizedBufferFaster(input: Buffer, output: Buffer) {
			var index = 0
			val arr = input.readByteArray()
			for (b in arr) {
				when (b) {
					9.toByte(), 10.toByte(), 13.toByte(), 32.toByte() -> {}
					else -> {
						arr[index] = b
						index++
					}
				}
			}
			output.write(arr, 0, index)
		}

	}

	private class Murmur2Hash : Hash {
		val value: Int

		constructor(value: String) {
			// Parsing as long then casting to int converts values gt int max value but lt uint max value
			// into negatives. I presume this is how the murmur2 code handles this.
			this.value = value.toLong().toInt()
		}

		constructor(value: Int) {
			this.value = value
		}

		override val stringValue get() = value.toString()
		override val type get() = "murmur2"

		override fun equals(other: Any?): Boolean {
			if (other !is Murmur2Hash) {
				return false
			}
			return value == other.value
		}

		override fun toString(): String = "murmur2: $value"
		override fun hashCode(): Int = value
	}

	override fun getHashingSource(delegate: Source): GeneralHashingSource = Murmur2GeneralHashingSource(delegate)
	override fun getHash(value: String): Hash = Murmur2Hash(value)
}